src/name_utils.rs

```diff
@@ -0,0 +1,399 @@
+/// (1 / number_of_sub_names) * (RESOURCE_NAME_MAX_LEN - KUBERNETES_HASH_MIN_LENGTH - number_of_sub_names)
+///
+/// If sub names are shorter than the basic length, these characters can be used by another (longer)
+/// sub name.
+///
+/// # Arguments
+///
+/// * `short_name` - The short name of the custom resource definition.
+/// * `cluster_name` - The name of the custom resource.
+/// * `role_name` - The role name of the custom resource.
+/// * `group_name` - Optional group name of the custom resource.
+/// * `node_name` - Optional node name if available (e.g. for pods).
+/// * `misc_name` - Optional miscellaneous identifiers (e.g. "data" for config maps).
+///
+pub fn build_resource_name(
+    short_name: &str,
+    cluster_name: &str,
+    role_name: &str,
+    group_name: Option<&str>,
+    node_name: Option<&str>,
+    misc_name: Option<&str>,
+) -> OperatorResult<String> {
+    // collect a vector of all sub names and their respective kind
+    let mut sub_names = vec![
+        // transform to alphanumeric and lowercase
+        to_alphanumeric_lowercase_not_empty(SubName::Short, short_name)?,
+        to_alphanumeric_lowercase_not_empty(SubName::Cluster, cluster_name)?,
+        to_alphanumeric_lowercase_not_empty(SubName::Role, role_name)?,
+    ];
+
+    if let Some(group) = group_name {
+        sub_names.push(to_alphanumeric_lowercase_not_empty(SubName::Group, group)?);
+    }
+
+    if let Some(node) = node_name {
+        sub_names.push(to_alphanumeric_lowercase_not_empty(SubName::Node, node)?);
+    };
+
+    if let Some(misc) = misc_name {
+        sub_names.push(to_alphanumeric_lowercase_not_empty(SubName::Misc, misc)?);
+    };
+
+    // Calculate the maximum number of available characters
+    // The sub names vector length equals the size of additional required dashes
+    let max_chars = RESOURCE_NAME_MAX_LEN - KUBERNETES_HASH_MIN_LENGTH - sub_names.len();
+    // The amount of characters receives without carryover (even distributed)
+    let selectable_chars = selectable_chars(&sub_names, max_chars);
+    // The `carryover` are left over characters that will not be used by some sub names
+    // and can be added to other longer sub names exceeding the amount of `selectable_chars`.
+    let carryover = max_chars - used_characters(&sub_names, selectable_chars);
+
+    Ok(build_name(
+        sub_names.as_slice(),
+        selectable_chars,
+        carryover,
+    ))
+}
+
+/// This splits the number of available characters into equal blocks for each sub name and
+/// returns its size.
+///
+/// # Arguments
+///
+/// * `sub_names` - A vector of available sub names.
+/// * `max_chars` - The maximum available characters:
+///                 `RESOURCE_NAME_MAX_LEN` - `KUBERNETES_HASH_MIN_LENGTH` - sub_names.len()
+///
+fn selectable_chars(sub_names: &[String], max_chars: usize) -> usize {
+    (1f32 / sub_names.len() as f32 * max_chars as f32).floor() as usize
```

Comment:
This is losing remainder characters. I think this is better:

```suggestion
        max_chars / sub_names.len() + max_chars % sub_names.len()
```

After this change, two tests need to be updated.

Reply:
Done.